misc/ls-extensions/modules/json-to-record-converter/src/main/java/io/ballerina/converters/JsonToRecordConverter.java

```diff
@@ -0,0 +1,400 @@
+import io.ballerina.converters.exception.ConverterException;
+import io.ballerina.converters.util.Constants;
+import io.ballerina.converters.util.ErrorMessages;
+import io.ballerina.converters.util.SchemaGenerator;
+import io.ballerina.tools.text.TextDocument;
+import io.ballerina.tools.text.TextDocuments;
+import io.swagger.v3.oas.models.Components;
+import io.swagger.v3.oas.models.OpenAPI;
+import io.swagger.v3.oas.models.media.ArraySchema;
+import io.swagger.v3.oas.models.media.ObjectSchema;
+import io.swagger.v3.oas.models.media.Schema;
+import io.swagger.v3.parser.OpenAPIV3Parser;
+import io.swagger.v3.parser.core.models.SwaggerParseResult;
+import org.apache.commons.lang3.StringUtils;
+import org.ballerinalang.formatter.core.Formatter;
+import org.ballerinalang.formatter.core.FormatterException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static io.ballerina.compiler.syntax.tree.NodeFactory.createBuiltinSimpleNameReferenceNode;
+import static io.ballerina.converters.util.ConverterUtils.convertOpenAPITypeToBallerina;
+import static io.ballerina.converters.util.ConverterUtils.escapeIdentifier;
+import static io.ballerina.converters.util.ConverterUtils.extractReferenceType;
+
+/**
+ * Implements functionality to convert Json (object or schema) to a ballerina record.
+ *
+ * The conversion is done by using openAPI schema as an intermediary.
+ *
+ * @since 2.0.0
+ */
+public class JsonToRecordConverter {
+
+    private JsonToRecordConverter() {
+        // not called
+    }
+
+    /**
+     * This method takes in a json string and returns the Ballerina record nodes.
+     *
+     * @param jsonString json string for the schema
+     * @return {@link ArrayList<TypeDefinitionNode>} arraylist of type nodes
+     * @throws IOException in case of Json parse error
+     * @throws ConverterException in case of invalid schema
+     */
+    public static String convert(String jsonString) throws IOException,
+            ConverterException, FormatterException {
+        ObjectMapper objectMapper = new ObjectMapper();
+        OpenAPI model;
+        JsonNode inputJson = objectMapper.readTree(jsonString);
+        if (inputJson.has("$schema")) {
+            model = parseJSONSchema(jsonString);
+        } else {
+            Map<String, Object> schema = SchemaGenerator.generate(inputJson);
+            String schemaJson = objectMapper.writeValueAsString(schema);
+            model = parseJSONSchema(schemaJson);
+        }
+        ArrayList<TypeDefinitionNode> typeDefinitionNodeList = generateRecords(model);
+        NodeList<ImportDeclarationNode> imports = AbstractNodeFactory.createEmptyNodeList();
+        NodeList<ModuleMemberDeclarationNode> moduleMembers = AbstractNodeFactory.createNodeList(
+                typeDefinitionNodeList.toArray(new TypeDefinitionNode[typeDefinitionNodeList.size()]));
```

Comment:
It might be better to define an empty array here instead of a pre-sized array.
```suggestion
                typeDefinitionNodeList.toArray(new TypeDefinitionNode[0]));
```

Reply:
Addressed in 333c06f